### PR TITLE
Add support for the Nokia N-Gage

### PR DESCRIPTION
**Pull request by mupfdev**
Add basic support for the Symbian S60v1/2 platform with a main focus on the Nokia N-Gage.

## Description
This port was created as part of the [N-Gage SDK project](https://github.com/ngagesdk).
It has been [proven to be functional](https://github.com/ngagesdk/wordle), however, maintenance in a separate fork is becoming increasingly cumbersome. I would therefore be very happy if it could become part of the upstream repository.

## Known issues
For the software renderer to work, clipping had to be disabled for the time being. This change only affects this platform. Unfortunately, I have not yet been able to find the cause of this problem.
This issue is mentioned in the respective README file.

### CODE REVIEW

File: src/SDL_log.c [hunk 1]
```diff
@@ -39,6 +39,11 @@
 #include <android/log.h>
 #endif
 
+#if defined(__NGAGE__)
+#undef va_copy
+#define va_copy(dst, src)   dst = src
```
Comment (slouken) [inline src/SDL_log.c:44]: We recently moved the definition of va_copy to SDL_internal.h, you could add the ngage version there.

File: src/stdlib/SDL_string.c [hunk 2]
```diff
@@ -28,6 +28,11 @@
 
 #include "SDL_stdinc.h"
 
+#if defined(__NGAGE__)
```
Comment (slouken) [inline src/stdlib/SDL_string.c:31]: You can remove this once you update SDL_internal.h

File: src/stdlib/SDL_malloc.c [hunk 3]
```diff
@@ -5304,6 +5304,11 @@ mspace_mallopt(int param_number, int value)
 
 #endif /* !HAVE_MALLOC */
 
+#ifdef __NGAGE__
```
Comment (slouken) [inline src/stdlib/SDL_malloc.c:5307]: Why is this needed? Does something need to be added to SDL_config_ngage.h instead?

**Comment by slouken** [comment]
I'm open to this, but I think we need to track down the clipping issue before we merge to the SDL repo.

File: src/stdlib/SDL_malloc.c (hunk 3 above)
Comment (mupfdev) [reply src/stdlib/SDL_malloc.c]: You're right; that's a remnant and doesn't belong there. I fixed it and the moved the `va_copy` definition as well. Thanks!

**Comment by mupfdev** [comment]
> I think you can just define HAVE_STDLIB_H in the header section below, and SDL will do the right thing.

I thought I had already tried that but apparently I was wrong. Works perfectly.

**Comment by mupfdev** [comment]
> I'm open to this, but I think we need to track down the clipping issue before we merge to the SDL repo.

Any tips on what to look for? I spent a few hours on this but couldn't determine the cause. The fact that I don't have a real debug option doesn't make it any easier, unfortunately.

**Comment by mupfdev** [comment]
I was able to isolate the cause of the clipping issue to `QueueCmdSetViewport()`:

https://github.com/libsdl-org/SDL/blob/f25a78eede5ef7da5f2d45ffc7f9f9e083abd265/src/render/SDL_render.c#L362-L363

`SDL_floor()` returns zero, although `renderer->viewport.w` and `renderer->viewport.h` have correct values. I keep looking.

**Comment by mupfdev** [comment]
Now everything works as intended.

**Comment by slouken** [comment]
Looks good! I merged everything into main, feel free to submit PRs to continue improving and maintaining the port.

File: src/thread/ngage/SDL_sysmutex.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    RMutex rmutex;
+    rmutex.SetHandle(mutex->handle);
+    rmutex.Wait();
+
+    return(0);
+}
+
+/* Unlock the mutex */
+int
+SDL_mutexV(SDL_mutex * mutex)
```
Comment (ccawley2011) [inline src/thread/ngage/SDL_sysmutex.cpp:92]:
```suggestion
SDL_UnlockMutex(SDL_mutex * mutex)
```

Also, there doesn't seem to be an implementation of `SDL_TryLockMutex`.

File: src/thread/ngage/SDL_sysmutex.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+#include "../../SDL_internal.h"
+
+/* An implementation of mutexes using semaphores */
```
Comment (ccawley2011) [inline src/thread/ngage/SDL_sysmutex.cpp:23]:
```suggestion
/* An implementation of mutexes using the Symbian API */
```

File: src/thread/ngage/SDL_syssem.cpp [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+#include "../../SDL_internal.h"
+
+/* An implementation of semaphores using mutexes and condition variables */
```
Comment (ccawley2011) [inline src/thread/ngage/SDL_syssem.cpp:23]:
```suggestion
/* An implementation of semaphores using the Symbian API */
```

File: src/thread/ngage/SDL_systhread.cpp [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+{
+    return;
+}
+
+/* WARNING: This function is really a last resort.
+ * Threads should be signaled and then exit by themselves.
+ * TerminateThread() doesn't perform stack and DLL cleanup.
+ */
+void
+SDL_SYS_KillThread(SDL_Thread *thread)
```
Comment (ccawley2011) [inline src/thread/ngage/SDL_systhread.cpp:136]: This seems to be a leftover from SDL 1.2.

File: src/video/ngage/SDL_ngageevents.cpp [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+            keycode = SDLK_0;
+            break;
+        case 0x2a:                // Asterisk
+            keycode = SDLK_ASTERISK;
+            break;
+        case EStdKeyHash:         // Hash
+            keycode = SDLK_SLASH;
+            break;
+        case EStdKeyDevice0:      // Left softkey
+            keycode = SDLK_F1;
```
Comment (ccawley2011) [inline src/video/ngage/SDL_ngageevents.cpp:119]: Would one of the newer SDL2 keycodes work better here?

**Comment by mupfdev** [comment]
Thank you!

File: src/video/ngage/SDL_ngageevents.cpp (hunk 8 above)
Comment (mupfdev) [reply src/video/ngage/SDL_ngageevents.cpp:119]: Yes, I'll try to choose them a bit more sense in the next PR.

File: src/thread/ngage/SDL_syssem.cpp (hunk 6 above)
Comment (mupfdev) [reply src/thread/ngage/SDL_syssem.cpp:23]: Updated.

File: src/thread/ngage/SDL_systhread.cpp (hunk 7 above)
Comment (mupfdev) [reply src/thread/ngage/SDL_systhread.cpp:136]: Removed it.

File: src/thread/ngage/SDL_sysmutex.cpp (hunk 4 above)
Comment (mupfdev) [reply src/thread/ngage/SDL_sysmutex.cpp:92]: I keep this unresolved for now until `SDL_TryLockMutex` has been implemented. I renamed `SDL_mutexV`.

File: src/main/ngage/SDL_ngage_main.cpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    if (NULL == newHeap)
+    {
+        ret = 3;
+        goto cleanup;
+    }
+    else
+    {
+        oldHeap = User::SwitchHeap(newHeap);
+        /* Call stdlib main */
+        ret = main(argc, argv);
```
[the remark this answers is not in the document]
Comment (mupfdev) [reply src/main/ngage/SDL_ngage_main.cpp:71]: Fixed it.
```C
SDL_SetMainReady();
ret = SDL_main(argc, argv);
```
